integration-tests/src/tests/client/features/in_memory_tries.rs

```diff
@@ -389,3 +474,160 @@ fn num_memtrie_roots(env: &TestEnv, client_id: usize, shard: ShardUId) -> Option
+        records.push(StateRecord::Account {
+            account_id: account.clone(),
+            account: Account::new(initial_balance, staked, CryptoHash::default(), 0),
+        });
+        records.push(StateRecord::AccessKey {
+            account_id: account.clone(),
+            public_key: create_user_test_signer(&account).public_key,
+            access_key: AccessKey::full_access(),
+        });
+        // The total supply must be correct to pass validation.
+        genesis_config.total_supply += initial_balance + staked;
+    }
+    let genesis = Genesis::new(genesis_config, GenesisRecords(records)).unwrap();
+    let chain_genesis = ChainGenesis::new(&genesis);
+
+    let mut clock = FakeClock::new(Utc::UNIX_EPOCH);
+    let stores = (0..NUM_VALIDATORS).map(|_| create_test_store()).collect::<Vec<_>>();
+    let mut env = TestEnv::builder(chain_genesis.clone())
+        .clients((0..NUM_VALIDATORS).map(|i| format!("account{}", i).parse().unwrap()).collect())
+        .stores(stores.clone())
+        .real_epoch_managers(&genesis.config)
+        .maybe_track_all_shards(track_all_shards)
+        .nightshade_runtimes_with_trie_config(
+            &genesis,
+            // Don't load any memtries.
+            (0..NUM_VALIDATORS).map(|_| TrieConfig::default()).collect(),
+        )
+        .build();
+
+    // Sanity check that we should have 4 block producers.
+    assert_eq!(
+        env.clients[0]
+            .epoch_manager
+            .get_epoch_block_producers_ordered(
+                &EpochId::default(),
+                &env.clients[0].chain.head().unwrap().last_block_hash
+            )
+            .unwrap()
+            .len(),
+        NUM_VALIDATORS
+    );
+
+    // Start the nodes from genesis, and then send transactions.
+    let mut nonces =
+        accounts.iter().map(|account| (account.clone(), 0)).collect::<HashMap<AccountId, u64>>();
+    let mut balances = accounts
+        .iter()
+        .map(|account| (account.clone(), initial_balance))
+        .collect::<HashMap<AccountId, u128>>();
+
+    run_chain_for_some_blocks_while_sending_money_around(
+        &mut clock,
+        &mut env,
+        &mut nonces,
+        &mut balances,
+        100,
+        track_all_shards,
+    );
+    // Assert that indeed no memtries are loaded.
+    for i in 0..NUM_VALIDATORS {
+        for shard_id in 0..4 {
+            assert_eq!(num_memtrie_roots(&env, i, ShardUId { version: 1, shard_id }), None);
+        }
+    }
+}
+
+#[test]
+fn test_in_memory_trie_consistency_with_state_sync_base_case_track_single_shard() {
+    test_in_memory_trie_consistency_with_state_sync_base_case(false);
+}
+
+#[test]
+fn test_in_memory_trie_consistency_with_state_sync_base_case_track_all_shards() {
+    test_in_memory_trie_consistency_with_state_sync_base_case(true);
+}
+
+trait TestEnvBuilderExt {
+    fn maybe_track_all_shards(self, track_all_shards: bool) -> Self;
```

Comment:
Why is this only for the `in_memory_tries` test? Eventually single shard tracking related tests should hopefully be more common?

Reply:
Ah yeah, I guess I should move that in there.